File: mf2/javascript/src/locale_direction.js
import { canonicalLocaleKey } from "./locale-key.js";
import * as data from "./locale_direction_data.js";

const sets = Object.fromEntries(Object.entries(data).map(([name, value]) => [name, new Set(value.trim().split(/\s+/))]));

export function localeIsLtr(locale) {
  let parts;
  try { parts = canonicalLocaleKey(locale).split("-"); } catch { return null; }
  const language = parts[0];
  let region = null;
  for (const part of parts.slice(1)) {
    if (/^[A-Z][a-z]{3}$/.test(part)) return sets.LTR_SCRIPTS.has(part) ? true : sets.RTL_SCRIPTS.has(part) ? false : null;
    if (/^(?:[A-Z]{2}|[0-9]{3})$/.test(part)) region = part;
  }
  if (region) {
    if (sets.LTR_REGION_OVERRIDES.has(`${language}-${region}`)) return true;
    if (sets.RTL_REGION_OVERRIDES.has(`${language}-${region}`)) return false;
  }
  if (language === "und") return null;
  return sets.LTR_LANGUAGES.has(language) ? true : sets.RTL_LANGUAGES.has(language) ? false : null;
}

File: mf2/python/src/mojito_mf2/_locale_direction.py
from ._locale_key import canonical_locale_key
from . import _locale_direction_data as data

_LTR_SCRIPTS = frozenset(data.LTR_SCRIPTS.split())
_RTL_SCRIPTS = frozenset(data.RTL_SCRIPTS.split())
_LTR_LANGUAGES = frozenset(data.LTR_LANGUAGES.split())
_RTL_LANGUAGES = frozenset(data.RTL_LANGUAGES.split())
_LTR_REGIONS = frozenset(data.LTR_REGION_OVERRIDES.split())
_RTL_REGIONS = frozenset(data.RTL_REGION_OVERRIDES.split())


def locale_is_ltr(locale: str) -> bool | None:
    try:
        parts = canonical_locale_key(locale).split("-")
    except (ValueError, TypeError):
        return None
    language, region = parts[0], None
    for part in parts[1:]:
        if len(part) == 4 and part.isalpha():
            return True if part in _LTR_SCRIPTS else False if part in _RTL_SCRIPTS else None
        if len(part) == 2 and part.isalpha() or len(part) == 3 and part.isdigit():
            region = part
    if region:
        key = f"{language}-{region}"
        if key in _LTR_REGIONS:
            return True
        if key in _RTL_REGIONS:
            return False
    if language == "und":
        return None
    return True if language in _LTR_LANGUAGES else False if language in _RTL_LANGUAGES else None
